core/grant/service.go

```diff
@@ -257,3 +268,102 @@ func (s *Service) expiredInActiveUserAccess(ctx context.Context, timeLimiter cha
 		}
 	}
 }
+
+type ImportAccessCriteria struct {
+	ProviderID   string `validate:"required"`
+	ResourceIDs  []string
+	ResouceTypes []string
+	ResourceURNs []string
+}
+
+func (s *Service) ImportAccess(ctx context.Context, criteria ImportAccessCriteria) ([]*domain.Grant, error) {
+	p, err := s.providerService.GetByID(ctx, criteria.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("getting provider details: %w", err)
+	}
+
+	listResourcesFilter := domain.ListResourcesFilter{
+		ProviderType: p.Type,
+		ProviderURN:  p.URN,
+	}
+	if criteria.ResourceIDs != nil {
+		listResourcesFilter.IDs = criteria.ResourceIDs
+	} else {
+		listResourcesFilter.ResourceTypes = criteria.ResouceTypes
+		listResourcesFilter.ResourceURNs = criteria.ResourceURNs
+	}
+	resources, err := s.resourceService.Find(ctx, listResourcesFilter)
+	if err != nil {
+		return nil, fmt.Errorf("getting resources: %w", err)
+	}
+
+	_, resourceAccess, err := s.providerService.ListAccess(ctx, criteria.ProviderID, resources)
+	if err != nil {
+		return nil, fmt.Errorf("fetching access from provider: %w", err)
+	}
+
+	resourcesMap := map[string]*domain.Resource{}
+	for _, r := range resources {
+		resourcesMap[r.URN] = r
+	}
+
+	grants, err := s.repo.List(ctx, domain.ListGrantsFilter{
+		ProviderTypes: []string{p.Type},
+		ProviderURNs:  []string{p.URN},
+		ResourceTypes: criteria.ResouceTypes,
+		ResourceURNs:  criteria.ResourceURNs,
+		Statuses:      []string{string(domain.GrantStatusActive)},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting active grants: %w", err)
+	}
+	// map[resourceURN]map[accounttype:accountId]map[role]grant
+	grantsMap := map[string]map[string]map[string]*domain.Grant{}
+	for _, g := range grants {
+		if grantsMap[g.Resource.URN] == nil {
+			grantsMap[g.Resource.URN] = map[string]map[string]*domain.Grant{}
+		}
+
+		userSignature := fmt.Sprintf("%s:%s", g.AccountType, g.AccountID)
+		if grantsMap[g.Resource.URN][userSignature] == nil {
+			grantsMap[g.Resource.URN][userSignature] = map[string]*domain.Grant{}
+		}
+
+		grantsMap[g.Resource.URN][userSignature][g.Role] = &g
+	}
+
+	importedGrants := []*domain.Grant{}
+	for rURN, access := range resourceAccess {
+		for _, ae := range access {
+			userSignature := fmt.Sprintf("%s:%s", ae.AccountType, ae.AccountID)
+			if grantsMap[rURN] != nil &&
+				grantsMap[rURN][userSignature] != nil &&
+				grantsMap[rURN][userSignature][ae.Permission] != nil {
+				continue // access already registered
+			}
+
+			r := resourcesMap[rURN]
```

Comment:
this could be done earlier under `for rURN, access := range resourceAccess {` loop